### PR TITLE
libckteec: better handle null size buffer in C_GetSlotList()

### PR DESCRIPTION
**Pull request by etienne-lms**
When caller of `C_GetSlotList()` provides a non-NULL buffer reference with a count value of zero, the TA gets invoked with a null buffer reference in the OP-TEE shared memory, hence it will return `CKR_OK` status since it legitimately only returns the count information.

This change makes ckteec library to return `CKR_BUFFER_TOO_SMALL` as per PKCS#11 specification.

This change also fixes `C_GetSlotList()` that, prior to this change, would have returned `CKR_ARGUMENTS_BAD` upon `NULL` buffer and non-zero count value whereas specification says:
> If pSlotList is NULL_PTR, (...). The contents of *pulCount on entry to C_GetSlotList has no meaning in this case (...).

Found the issue from https://github.com/OP-TEE/optee_os/pull/3644#discussion_r387592239.

### CODE REVIEW

**Comment by etienne-lms** [comment]
xtest update for the specific case through https://github.com/OP-TEE/optee_test/pull/400.

File: libckteec/src/pkcs11_token.c [hunk 1]
```diff
@@ -83,6 +92,19 @@ CK_RV ck_slot_get_list(CK_BBOOL present,
...
 	if (rv == CKR_OK || rv == CKR_BUFFER_TOO_SMALL) {
+		/*
+		 * When @slot != NULL and @count[0] == 0, library shall
+		 * return CKR_BUFFER_TOO_SMALL if provided buffer is too
+		 * small whereas TA would have returned CKR_OK since
+		 * allocating a zero sized shm instance results in a NULL
+		 * shm buffer reference.
+		 */
+		if (size && slots && !client_count) {
+			assert(rv == CKR_OK);
```
Comment (jenswi-linaro) [inline libckteec/src/pkcs11_token.c]: I'm not convinced that we should have this assert().

Comment (etienne-lms) [reply libckteec/src/pkcs11_token.c]: True, I can remove.
(will do)

**Comment by jenswi-linaro** [comment]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by etienne-lms** [comment]
Tag applied.
Failure in IBART seems not related to these changes:
`23805:  error: RPC failed; curl 56 GnuTLS recv error (-110): The TLS connection was non-properly terminated.`

**Pull request merged**